crates/voicevox_core/src/user_dict/word.rs

```diff
@@ -0,0 +1,225 @@
+            return Err(Error::InvalidWord(
+                "発音は有効なカタカナである必要があります。".to_string(),
+            ));
+        }
+        let sutegana = ['ァ', 'ィ', 'ゥ', 'ェ', 'ォ', 'ャ', 'ュ', 'ョ', 'ヮ', 'ッ'];
+
+        let pronunciation_chars = pronunciation.chars().collect::<Vec<_>>();
+
+        for i in 0..pronunciation_chars.len() {
+            // 「キャット」のように、捨て仮名が連続する可能性が考えられるので、
+            // 「ッ」に関しては「ッ」そのものが連続している場合と、「ッ」の後にほかの捨て仮名が連続する場合のみ無効とする
+            if sutegana.contains(pronunciation_chars.get(i).unwrap())
+                && i < pronunciation_chars.len() - 1
+                && (sutegana[..sutegana.len() - 1]
+                    .contains(pronunciation_chars.get(i + 1).unwrap())
+                    || (pronunciation_chars.get(i).unwrap() == &'ッ'
+                        && sutegana.contains(pronunciation_chars.get(i + 1).unwrap())))
+            {
+                return Err(Error::InvalidWord(
+                    "無効な発音です。(捨て仮名の連続)".to_string(),
+                ));
+            }
+
+            if pronunciation_chars.get(i).unwrap() == &'ヮ'
+                && i != 0
+                && !['ク', 'グ'].contains(pronunciation_chars.get(i - 1).unwrap())
+            {
+                return Err(Error::InvalidWord(
+                    "無効な発音です。(「くゎ」「ぐゎ」以外の「ゎ」の使用)".to_string(),
+                ));
+            }
+        }
+        Ok(())
+    }
+
+    /// カタカナの発音からモーラ数を計算する。
+    fn calculate_mora_count(pronunciation: &str, accent_type: usize) -> Result<usize> {
+        // 元実装：https://github.com/VOICEVOX/voicevox_engine/blob/39747666aa0895699e188f3fd03a0f448c9cf746/voicevox_engine/model.py#L212-L236
+        let rule_others =
+            r#"[イ][ェ]|[ヴ][ャュョ]|[トド][ゥ]|[テデ][ィャュョ]|[デ][ェ]|[クグ][ヮ]"#;
+        let rule_line_i = r#"[キシチニヒミリギジビピ][ェャュョ]"#;
+        let rule_line_u = r#"[ツフヴ][ァ]|[ウスツフヴズ][ィ]|[ウツフヴ][ェォ]"#;
+        let rule_one_mora = r#"[ァ-ヴー]"#;
+
+        let mora_count = regex::Regex::new(&format!(
+            r#"(?:{}|{}|{}|{})"#,
+            rule_others, rule_line_i, rule_line_u, rule_one_mora
+        ))
+        .unwrap()
+        .find_iter(pronunciation)
+        .count();
+
+        if accent_type > mora_count {
+            return Err(Error::InvalidWord(format!(
+                "誤ったアクセント型です({})。 expect: 0 <= accent_type <= {}",
+                accent_type, mora_count
+            )));
+        }
+
+        Ok(mora_count)
+    }
+
+    /// ASCII文字を全角文字に変換する。
+    fn to_zenkaku(surface: &str) -> String {
+        // 元実装：https://github.com/VOICEVOX/voicevox/blob/69898f5dd001d28d4de355a25766acb0e0833ec2/src/components/DictionaryManageDialog.vue#L379-L387
+        let mut result = String::new();
+        for c in surface.chars() {
+            let i = c as u32;
+            result.push(if (0x21..=0x7e).contains(&i) {
+                char::from_u32(0xfee0 + i).unwrap_or(c)
+            } else if i == 0x20 {
+                '　'
+            } else {
+                c
+            });
+        }
+        result
```

Comment:
`<String as FromIterator<char>>`を利用すればすっきりすると思います。

```suggestion
        surface
            .chars()
            .map(|c| match u32::from(c) {
                i @ 0x21..=0x7e => char::from_u32(0xfee0 + i).unwrap_or(c),
                0x20 => '　',
                _ => c,
            })
            .collect()
```

あと`DictionaryManageDialog`の実装と比較してなのですが、`0x20`って`\p{Z}`より狭いような...?